[svg] marker properties should allow any URI

graouts opened the pull request:
#### adab1550880a6b4d78ec26f85fbfd3105e34b60b
<pre>
[svg] marker properties should allow any URI
<a href="https://bugs.webkit.org/show_bug.cgi?id=252163">https://bugs.webkit.org/show_bug.cgi?id=252163</a>

Reviewed by Said Abou-Hallawa.

Resolve the fragment of the URL passed to the marker properties at use time rather than at parse-time.

We remove a non-WPT test that was incorrect in its assumptions and testing functionality already covered
by WPT.

* LayoutTests/imported/w3c/web-platform-tests/svg/painting/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-end-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-mid-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-start-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/svg/css/computed-style-url-serializing-expected.txt: Removed.
* LayoutTests/svg/css/computed-style-url-serializing.html: Removed.
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::markerValue):
(WebCore::ComputedStyleExtractor::svgPropertyValue):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGPaintingFeatures):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSVGURIReference):

Canonical link: <a href="https://commits.webkit.org/260299@main">https://commits.webkit.org/260299@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6389061fc9c64b0c34ce310a3657a40790209b70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107771 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16827 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40668 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116916 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116297 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18241 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8143 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99951 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113528 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13772 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41463 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95679 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28611 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9791 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29959 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10478 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6849 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15945 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49543 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12041 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3864 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
```diff
@@ -84,6 +85,16 @@ Ref<CSSValue> ComputedStyleExtractor::adjustSVGPaint(SVGPaintType paintType, con
     return color;
 }
 
+static RefPtr<CSSValue> markerValue(const String& marker, const Element* element)
```
I think this should be named `svgMarkerValue()`. The marker is used for other things, e.g. `DocumentMarker`

graouts replied:
Done in latest commit.

shallawa commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1591,14 +1591,12 @@ inline float BuilderConverter::convertOpacity(BuilderState&, const CSSValue& val
 }
 
-inline String BuilderConverter::convertSVGURIReference(BuilderState& builderState, const CSSValue& value)
+inline String BuilderConverter::convertSVGURIReference(BuilderState&, const CSSValue& value)
```
Why do we still have to pass `BuilderState` to this function and the caller functions like `applyValueMarkerStart()`? I think it should be removed.

graouts replied:
I believe the call sites are generated and probably always provide the `BuilderState`. I'll take a look to see if that can be changed however.

graouts replied:
Yes, I've confirmed all those call sites come from `StyleBuilderGenerated`. So I'll leave this as-is.

shallawa commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -265,19 +265,19 @@ bool SVGResources::buildCachedResources(const RenderElement& renderer, const Ren
     if (markerTags().contains(tagName) && svgStyle.hasMarkers()) {
-        AtomString markerStartId(svgStyle.markerStartResource());
+        auto markerStartId = SVGURIReference::fragmentIdentifierFromIRIString(svgStyle.markerStartResource(), document);
         if (setMarkerStart(getRenderSVGResourceById<RenderSVGResourceMarker>(document, markerStartId)))
```
Should not this be refactored into a helper function or a template specialization instead of repeating the same call?

```
template<>
Renderer* getRenderSVGResourceById<RenderSVGResourceMarker>(Document& document, const AtomString& id)
{
    auto markerId = SVGURIReference::fragmentIdentifierFromIRIString(id, document);
    RenderSVGResource* container = getRenderSVGResourceContainerById(document, markerId);
    if (is<Renderer>(container))
        return downcast<Renderer>(container);
    return nullptr;
}
```

graouts replied:
I wrote this in the updated patch.

```c++
auto buildCachedMarkerResource = [&](const String& markerResource, bool (SVGResources::*setMarker)(RenderSVGResourceMarker*)) {
    auto markerId = SVGURIReference::fragmentIdentifierFromIRIString(markerResource, document);
    if ((this->*setMarker)(getRenderSVGResourceById<RenderSVGResourceMarker>(document, markerId)))
        foundResources = true;
    else
        registerPendingResource(extensions, markerId, element);
};
buildCachedMarkerResource(svgStyle.markerStartResource(), &SVGResources::setMarkerStart);
buildCachedMarkerResource(svgStyle.markerMidResource(), &SVGResources::setMarkerMid);
buildCachedMarkerResource(svgStyle.markerEndResource(), &SVGResources::setMarkerEnd);
```

graouts committed:
[svg] marker properties should allow any URI
https://bugs.webkit.org/show_bug.cgi?id=252163

Reviewed by Said Abou-Hallawa.

Resolve the fragment of the URL passed to the marker properties at use time rather than at parse-time.

We remove a non-WPT test that was incorrect in its assumptions and testing functionality already covered
by WPT.

* LayoutTests/imported/w3c/web-platform-tests/svg/painting/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-end-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-mid-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/parsing/marker-start-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/svg/css/computed-style-url-serializing-expected.txt: Removed.
* LayoutTests/svg/css/computed-style-url-serializing.html: Removed.
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::markerValue):
(WebCore::ComputedStyleExtractor::svgPropertyValue):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGPaintingFeatures):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSVGURIReference):

Canonical link: https://commits.webkit.org/260299@main
